simulation/dashboard.py

```diff
@@ -0,0 +1,290 @@
+import plotly_express as px
+from dash import html, dcc, dash
+import numpy as np
+import pandas as pd
+import json
+
+
+def get_data():
+    with open("./outputs/raw.json") as f:
+        data = json.load(f)
+
+    messages = {
+        "run_id": [],
+        "status": [],
+        "failure_rate": [],
+        "observed_failure_rate": [],
+        "type": [],
+        "emitter_time": [],
+        "receiver_time": [],
+        "emitter_id": [],
+        "receiver_id": [],
+        "delivered": [],
+    }
+
+    for i, run in enumerate(data):
+        for message in run["messages"]:
+            messages["run_id"].append(i)
+            messages["status"].append(run["status"])
+            messages["failure_rate"].append(run["failureRate"])
+            messages["observed_failure_rate"].append(run["observedFailureRate"])
+            messages["type"].append(message["type"])
+            messages["emitter_time"].append(message["emissionTime"])
+            messages["receiver_time"].append(message["receptionTime"])
+            messages["emitter_id"].append(message["emitterId"])
+            messages["receiver_id"].append(message["receiverId"])
+            messages["delivered"].append(message["delivered"])
+
+    df = pd.DataFrame(messages)
+    df["latency"] = (df["receiver_time"] - df["emitter_time"]).apply(
+        lambda x: max(0, x)
+    )
+
+    df["simulation_length"] = df["receiver_time"]
+
+    maxs = {}
+    ids = [i for i in pd.unique(df["run_id"])]
+    for i in ids:
+        maxs[i] = df[df["run_id"] == i]["receiver_time"].max()
+
+    def simlen(x):
+        x["simulation_length"] = maxs[x["run_id"]]
+        return x
+
+    df.apply(simlen, axis=1)
+    return df
+
+
+if __name__ == "__main__":
+    data = get_data()
```

Comment:
The script will crash if there is no data. You should add some checks to properly exit when this happens